[NFC] Introduce isType and getDeclType

Azoy opened the pull request:
I've noticed that there are a lot of cases where we dance around types by grabbing their nominal type and comparing that to some stdlib decl. Let's clean this up a little bit and introduce `isType` so that instead of dancing around this:
```cpp
if (type->getAnyNominal() == C.getIntDecl())
```
it becomes:
```cpp
if (type->isInt())
```
Note: this is only applied to those stdlib types defined in `KnownStdlibTypes`.

I've also introduced `getDeclType` which is shorthand for e.g. `C.getBoolDecl()->getDeclaredInterfaceType()` == `C.getBoolType()`.

@slavapestov @jrose-apple (Sorry if this is a lot to review!)

xwu commented:
For consistency, shouldn't these be called `isIntType()`, etc.?

Azoy commented:
> For consistency, shouldn't these be called `isIntType()`, etc.?

They can be, but there was already a `isBool` and `isVoid` for precedent. It also seems kind of redundant because you're saying type twice in a sentence. "If this type is equal to Int type." It doesn't really matter to me though at the end of the day which ever way this goes.

jrose-apple commented on lib/AST/ASTContext.cpp:
```diff
@@ -670,8 +670,7 @@ FuncDecl *ASTContext::getPlusFunctionOnRangeReplaceableCollection() const {
         if (Req.getKind() == RequirementKind::Conformance &&
-              Req.getSecondType()->getNominalOrBoundGenericNominal() ==
-            getRangeReplaceableCollectionDecl()) {
+              Req.getSecondType()->isRangeReplaceableCollection()) {
```
This has me thinking that the "Type" suffix might be valuable, to not imply that this is doing a compatibility check. `Array<Int>` "is a RangeReplaceableCollection" in some sense.

Azoy replied:
There are a couple of protocols listed in KnownStdlibTypes that could be moved over to KnownProtocols, that way they would still have to compare against the decl. We could introduce a isProtocolType for those.

jrose-apple commented on lib/AST/Type.cpp:
```diff
@@ -556,19 +556,18 @@ bool TypeBase::hasTypeRepr() const {
+    if (isa<DECL_CLASS>(generic)) \
+      return getASTContext().get##NAME##Decl() == generic; \
+  /** Edge case for Void, could be () too */ \
+  if (#NAME == "Void") \
```
You're doing a pointer comparison on the string literals, which isn't really guaranteed to work. Convert one of them to StringRef (or llvm::StringLiteral) to get the behavior you want.

(And in practice, you might as well do this first and have the other part in an `else` branch.)

jrose-apple replied:
Also, the way this logic is set up, you're still never going to get a TypeAliasDecl to return `true` here (which is probably a good thing for types). Maybe it'd be better to just separate typealiases out specially on TypeBase, and then `isVoid` can go back to being manually implemented.

Azoy replied:
I want to clean up KnownStdlibTypes soon, so I can do things like separate typealiases out, but I'll save those cleanups for a future patch. In any case, I'd like it if you reviewed the new implementation of this.

jrose-apple commented on lib/PrintAsObjC/DeclAndTypePrinter.cpp:
```diff
@@ -1137,22 +1137,19 @@ class DeclAndTypePrinter::Implementation
+      if (nominal && isa<StructDecl>(nominal) &&
+          (copyTy->isArray() || copyTy->isDictionary() || copyTy->isSet() ||
+          copyTy->isString() ||
+          (!getKnownTypeInfo(nominal) && getObjCBridgedClass(nominal)))) {
```
Please put this back as a nested `if` to not mix `&&` and `||` even further. (The paren alignment got messed up and that made me double-take as it was.)

Azoy replied:
oops I forgot this one, thanks!

jrose-apple commented on lib/SILGen/SILGen.cpp:
```diff
@@ -452,17 +452,13 @@ SILFunction *SILGenModule::emitTopLevelFunction(SILLocation Loc) {
-    Int32Ty = Int32Decl->getDeclaredInterfaceType()->getCanonicalType();
-  } else {
+  CanType Int32Ty = CanType(C.getInt32Type());
+  if (Int32Ty)
```
`if (!Int32Ty)`

jrose-apple commented on lib/Sema/CSApply.cpp:
```diff
@@ -264,14 +264,12 @@ static Expr *buildDynamicMemberLookupIndexExpr(StringRef name, SourceLoc loc,
-  nameExpr->setType(stringType);
+  nameExpr->setBuiltinInitializer(ctx.getStringBuiltinInitDecl(
+                                    ctx.getStringDecl()));
+  nameExpr->setType(ctx.getStringType());
```
I don't think this ended up cleaner. It should still be okay to say things the long way.

Azoy replied:
In this specific example, or across the board?

jrose-apple replied:
In this specific example. None of the others jumped out at me like this where repeated use seemed less clear, though I do hope it doesn't have any noticeable effect on compile time.

jrose-apple commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -7165,8 +7165,7 @@ ConstraintSystem::simplifyRestrictedConstraintImpl(
     type1 = simplifyType(type1);
-    if (isAnyHashableType(
-            type1->getRValueType()->lookThroughAllOptionalTypes())) {
+    if (type1->getRValueType()->lookThroughAllOptionalTypes()->isAnyHashable()) {
```
Just barely over the 80-column limit.

jrose-apple reviewed (commented):
I wasn't sure at first but looking through the diff shows that (a) an awful lot of code has gotten cleaner, and (b) we were already writing helpers to do basically this anyway. I do have comments but the general approach seems reasonable.

LucianoPAlmeida commented:
Would make sense or be useful to add something like [this isStdlibType](https://github.com/apple/swift/blob/80ef4c66c270b743461b582ed040dc6f5edbf863/lib/Sema/CSDiagnostics.cpp#L681) here too?

Azoy commented:
> Would make sense or be useful to add something like [this isStdlibType](https://github.com/apple/swift/blob/80ef4c66c270b743461b582ed040dc6f5edbf863/lib/Sema/CSDiagnostics.cpp#L681) here too?

It might make sense, maybe not in this patch though.

Azoy commented:
Will fix conflict soon.

jrose-apple commented on lib/AST/Type.cpp:
```diff
@@ -556,18 +556,26 @@ bool TypeBase::hasTypeRepr() const {
 bool TypeBase::isVoid() {
-  if (auto TT = getAs<TupleType>())
-    return TT->getNumElements() == 0;
+  if (auto tuple = dyn_cast<TupleType>(this)) {
```
This should continue to look through type sugar (i.e. the previous implementation was correct).

jrose-apple reviewed (commented):
Last comment, I think!

Azoy commented:
@jrose-apple alright, should be good to go!

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

Azoy commented:
Are the failures related?

futurejones commented:
Can one of the admins verify this patch?

CodaFi commented:
We got this partially done in the linked PR. Do you have the time to rebase this and push on with the rest?

Azoy commented:
@CodaFi I can update this weekend if that's fine.

Azoy committed:
Introduce isType and getDeclType

Accidentally forgot a negation

address review notes

Address more of Jordan's feedback

Azoy commented:
@CodaFi sorry this is a little later than last weekend, but I've rebased and fixed the issue upthread. Please take a look!

CodaFi commented on lib/AST/ASTContext.cpp:
```diff
@@ -738,22 +734,28 @@ FuncDecl *ASTContext::getSequenceMakeIterator() const {
+\
+Type ASTContext::get##NAME##Type() const { \
+  if (!get##NAME##Decl()) \
+    return Type(); \
```
This is the one thing I'm worried about. Clients of these accessors are historically terrible at handling the case where the lookup fails. That's why some of the TypeChecker entrypoints try to detect this, emit a diagnostic, and map the null type to an `ErrorType`. There shouldn't be any clients of this function before Sema, so maybe we should just look into doing that everywhere.

Azoy replied:
Slava pointed out in a previous commit that I did that perhaps if these accessors fail to just fail loudly because other pipelines of the compiler aren't great at handling whether or not they failed. Would that ease your concern, or should we still diagnose and provide an error type still?

CodaFi commented:
@swift-ci test

Azoy committed:
BuilderTransform: Use optional declared type

Azoy commented:
@CodaFi I've fixed the test issues related to function builders!

CodaFi commented:
@swift-ci test

CodaFi commented:
@swift-ci test

Azoy commented:
Ok, I’m going to take another look at the failing tests and resolve them.

varungandhi-apple commented on lib/IRGen/GenClangType.cpp:
```diff
@@ -303,12 +303,12 @@ ClangTypeConverter::reverseBuiltinTypeMapping(IRGenModule &IGM,
     // the NSInteger and NSUInteger typedefs, so map them back to those typedefs
     // if they're available, to ensure we get consistent ObjC @encode strings.
-    if (swiftType->getAnyNominal() == IGM.Context.getIntDecl()) {
+    if (swiftType->isInt()) {
```
For changes in this file, please also update the corresponding code in `ClangTypeConverter.cpp`, so that they don't get out of sync. 🙂 See my comment [here](https://github.com/apple/swift/pull/30630#discussion_r405057467) in case you're wondering why there is this duplication.

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))
